common/txmgr/address_state.go

```diff
@@ -188,32 +189,124 @@ func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) AddTxT
+	as.Lock()
+	defer as.Unlock()
+
+	tx, ok := as.confirmedMissingReceiptTxs[txID]
+	if !ok || tx == nil {
+		return fmt.Errorf("move_confirmed_missing_receipt_to_unconfirmed: no confirmed_missing_receipt transaction with ID %d", txID)
+	}
+
+	tx.State = TxUnconfirmed
+	as.unconfirmedTxs[tx.ID] = tx
+	delete(as.confirmedMissingReceiptTxs, tx.ID)
+
 	return nil
 }
 
 // MoveInProgressToUnconfirmed moves the in-progress transaction to the unconfirmed state.
+// It returns an error if there is no transaction in progress.
+// It returns an error if the transaction in progress does not have an attempt with the given ID.
 func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveInProgressToUnconfirmed(
-	etx txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
-	txAttempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	txError null.String, broadcastAt time.Time, initialBroadcastAt time.Time,
+	txAttemptID int64,
 ) error {
+	as.Lock()
+	defer as.Unlock()
+
+	tx := as.inprogressTx
+	if tx == nil {
+		return fmt.Errorf("move_in_progress_to_unconfirmed: no transaction in progress")
+	}
+
+	tx.State = TxUnconfirmed
+	tx.Error = txError
+	tx.BroadcastAt = &broadcastAt
+	tx.InitialBroadcastAt = &initialBroadcastAt
+
+	found := false
+	for i := 0; i < len(tx.TxAttempts); i++ {
+		txAttempt := tx.TxAttempts[i]
+		if txAttempt.ID == txAttemptID {
+			txAttempt.State = txmgrtypes.TxAttemptBroadcast
+			tx.TxAttempts[i] = txAttempt
+			found = true
+			break
+		}
+	}
+	if !found {
+		return fmt.Errorf("move_in_progress_to_unconfirmed: transaction in progress does not have an attempt with ID %d", txAttemptID)
+	}
+
+	as.unconfirmedTxs[tx.ID] = tx
+	as.inprogressTx = nil
+
 	return nil
 }
 
 // MoveUnconfirmedToConfirmed moves the unconfirmed transaction to the confirmed state.
+// It returns an error if there is no unconfirmed transaction with the given ID.
+// It returns an error if there is no attempt with the given receipt.
 func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveUnconfirmedToConfirmed(
 	receipt txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
 ) error {
+	as.Lock()
+	defer as.Unlock()
+
+	txAttempt, ok := as.attemptHashToTxAttempt[receipt.GetTxHash()]
+	if !ok {
+		return fmt.Errorf("move_unconfirmed_to_confirmed: no unconfirmed transaction with receipt %v", receipt)
+	}
+	// TODO(jtw): not sure how to set blocknumber, transactionindex, and receipt on conflict
+	txAttempt.Receipts = []txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH]{receipt}
+	txAttempt.State = txmgrtypes.TxAttemptBroadcast
+	if txAttempt.BroadcastBeforeBlockNum == nil {
+		blockNum := receipt.GetBlockNumber().Int64()
+		txAttempt.BroadcastBeforeBlockNum = &blockNum
+	}
+	tx, ok := as.unconfirmedTxs[txAttempt.TxID]
+	if !ok {
+		// TODO: WHAT SHOULD WE DO HERE?
+		// THIS WOULD BE A BIG BUG
+		return fmt.Errorf("move_unconfirmed_to_confirmed: no unconfirmed transaction with ID %d", txAttempt.TxID)
+	}
```

Comment:
We would maybe want to do this check towards the top. Either before or after checking if we can find the attempt. Since everything is in-memory now, we probably want to return any errors at the top of the function only. Otherwise, we'd be making changes to the state we wouldn't want like updating the attempt here.

Reply:
I noticed we're still making an update to the attempt before this check. Do we still want to set the receipt in the attempt even if there's a chance we can't find an associated tx and return an error?